File: Sources/VoiceFlowCore/TextJoiner.swift
import Foundation

/// What sits immediately before the insertion point in the target app.
public enum CursorContext: Equatable {
    /// The target app doesn't expose its text through Accessibility.
    case unavailable
    /// The cursor is at the very beginning of the text.
    case atStart
    case character(Character)
}

/// Decides whether a space must be inserted before a new transcription so
/// consecutive dictations don't run together ("siamo!Adesso").
public enum TextJoiner {
    public static func prefix(for text: String, precededBy context: CursorContext) -> String {
        guard let first = text.first, startsWord(first) else { return "" }
        switch context {
        case .unavailable, .atStart:
            return ""
        case .character(let previous):
            return needsSpace(after: previous) ? " " : ""
        }
    }

    private static func startsWord(_ c: Character) -> Bool {
        c.isLetter || c.isNumber
    }

    private static func needsSpace(after previous: Character) -> Bool {
        if previous.isWhitespace || previous.isNewline { return false }
        // Openers attach to what follows them.
        if "([{\"'«‘“".contains(previous) { return false }
        return true
    }
}

File: windows/VoiceFlow.Core/TextJoiner.cs
using System.Text;

namespace VoiceFlow.Core;

/// <summary>
/// Decides whether a space must be inserted before a new transcription so
/// consecutive dictations don't run together ("siamo!Adesso").
/// </summary>
public static class TextJoiner
{
    // Openers attach to what follows them; the straight apostrophe counts too, since it also opens an elided word in Italian ("l'amico").
    private const string Openers = "([{\"'«‘“";

    public static string Prefix(string text, CursorContext precededBy)
    {
        if (string.IsNullOrEmpty(text) || !StartsWord(text))
        {
            return "";
        }

        return precededBy switch
        {
            CursorContext.Character character => NeedsSpace(character.Value) ? " " : "",
            _ => "",
        };
    }

    private static bool StartsWord(string text)
    {
        var rune = Rune.GetRuneAt(text, 0);
        return Rune.IsLetterOrDigit(rune);
    }

    private static bool NeedsSpace(char previous)
    {
        if (char.IsWhiteSpace(previous))
        {
            return false;
        }

        return !Openers.Contains(previous);
    }
}
